### PR TITLE
DAOS-15240 cq: Add a GitHub actions unit test stage.

### PR DESCRIPTION
**Pull request by ashleypittman**
### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/control/run_go_tests.sh [hunk 1]
```diff
@@ -117,50 +117,6 @@ function setup_environment()
 	fi
 }
 
-function emit_junit_failure()
```
Comment (mjmac) [inline src/control/run_go_tests.sh:120]: What's the motivation for removing this function? If it's not currently needed, then we could adjust the logic so that it's not called for the current CI environment. I'm hesitant to just toss it altogether though, as this precise format was found to work with Jenkins' JUnit parser via trial and error.

File: src/control/run_go_tests.sh [hunk 2]
```diff
@@ -117,50 +117,6 @@ function setup_environment()
...
-<?xml version="1.0" encoding="UTF-8" ?>
-<testsuites>
-  <testsuite tests="1" failures="1" errors="0" skipped="0" >
-    ${teststr}
-  </testsuite>
-</testsuites>
-EOF
-}
-
-function check_formatting()
```
Comment (mjmac) [inline src/control/run_go_tests.sh:143]: I think this function is still useful for developers who run the script locally (anyone who touches Go code is encouraged to run it prior to pushing a PR). ... After reading the rest of this PR, I see that you extracted it to a standalone script. I guess maybe we could also add that script to the pre-commit git hooks?

Comment (ashleypittman) [reply src/control/run_go_tests.sh:143]: The logic with this file was to move the check to where it ran outside of the Intel CI infrastructure so that the results were available to all.  This just seemed an easy thing to include and it doesn't natively support JUnit - which works with Jenkins but not with GitHub actions unfortunately.

I'm torn between wanting to run things quicker with the results available to all in GitHub actions or with the better reporting but taking longer and visible to fewer people.

It's easy enough to back out this change, I just saw it as something easy to move but hadn't considered your point about having a single entry point for developers - that's nice that you've got that process that works.

Comment (mjmac) [reply src/control/run_go_tests.sh:143]: I had completely forgotten about the gofmt githook that I added in December of 2022. I have no problem with removing this from the test runner script and into its own GHA-compatible script. At some point, it would be nice to make the GHA gofmt check work like the other linters, so that it shows the required updates inline as comments, but that could be done later.

Comment (ashleypittman) [reply src/control/run_go_tests.sh:143]: To make annotations work then you need the source line, gofmt gives you a diff but it's not easy to go from there to the line.

You've just reminded me of something though, in Jenkins checkpatch we used to have a process where we had scripts that could "fix" the source tree by patching them as gofmt has an option to do, then our code would run "git diff" to see where the changes were and if they were in a line modified by the PR then they'd annotate them as comments to the PR.  This was used by codespell and a gofmt hook, but was also really flaky and used to break a lot so I removed it when I moved codespell to GitHub actions.

File: utils/utest.yaml [hunk 3]
```diff
@@ -2,8 +2,10 @@
 #
 #  SPDX-License-Identifier: BSD-2-Clause-Patent
 - name: include checks
+  memcheck: False
+  gha: True
   tests:
-    - cmd: ["utils/include_test.py"]
+    - cmd: ["utils/unit/include_test.py"]
```
Comment (jolivier23) [inline utils/utest.yaml:8]: I presume the idea is to eventually run more of them?

Comment (jolivier23) [reply utils/utest.yaml:8]: Maybe this could be more generic...such as 
tags: gha

and using --tags gha

**Review by jolivier23 — APPROVED** [review]
Ideally, we would move all unit testing to gha

File: utils/utest.yaml (hunk 3 above)
Comment (ashleypittman) [reply utils/utest.yaml:8]: @daltonbohning could comment on this.  There already is a `--suite` option but it does string matching rather than using tags so it's not much use IMO.

**Comment by ashleypittman** [comment]
> Ideally, we would move all unit testing to gha

I agree, that's the whole point of making a start to see how far we can go with that.

Jenkins has better reporting of junit results and can handle valgrind xml files which seems to be entirely missing from GitHub actions so in some ways it would be a step backwards.

File: utils/utest.yaml (hunk 3 above)
Comment (daltonbohning) [reply utils/utest.yaml:8]: I recently added this one, which might belong here ( I need to think about it some more)
https://github.com/daos-stack/daos/blob/40f9ffaa9dbc97f68be6450cdf9ec58c371c6c1e/utils/utest.yaml#L185-L189
And I suspect there are more we could group together.

And at least for python tests, we might consider running `pytest` against a directory of unit tests.

**Comment by daltonbohning** [comment]
> > Ideally, we would move all unit testing to gha
> 
> I agree, that's the whole point of making a start to see how far we can go with that.
> 
> Jenkins has better reporting of junit results and can handle valgrind xml files which seems to be entirely missing from GitHub actions so in some ways it would be a step backwards.

Do you know if valgrind can report in a format other than XML? XML has been losing popularity and generally has less support for parsing and displaying than, for example, json.

File: .github/workflows/bash_unit_testing.yml [hunk 4]
```diff
@@ -1,7 +1,6 @@
 name: bash Unit Testing
 
 on:
-  push:
```
Comment (daltonbohning) [inline .github/workflows/bash_unit_testing.yml:4]: Does removing this mean the action only runs once when the pull request is created?

Comment (daltonbohning) [inline .github/workflows/bash_unit_testing.yml:4]: Does removing this mean the action only runs once when the pull request is created?

File: .github/workflows/bash_unit_testing.yml [hunk 5]
```diff
@@ -15,6 +14,7 @@ defaults:
 jobs:
   Test-gha-functions:
     name: Tests in ci/gha_functions.sh
+    if: github.repository == 'daos-stack/daos'
```
Comment (daltonbohning) [inline .github/workflows/bash_unit_testing.yml:17]: Is this related to pushing PRs from forks? Why would the repo not be `daos-stack/daos`?

File: .github/workflows/bash_unit_testing.yml (hunk 4 above)
Comment (ashleypittman) [reply .github/workflows/bash_unit_testing.yml:4]: Yes.  As it stands this runs whenever there's a commit, so for example when you push a branch without creating a PR or more commonly when a PR is merged then it runs against master, neither of which is helpful or done by any other of our checks.

File: .github/workflows/bash_unit_testing.yml (hunk 5 above)
Comment (ashleypittman) [reply .github/workflows/bash_unit_testing.yml:17]: Kind of, it's related to pushing PRs to forks.  Doing anything in a fork I get lots of email about unable to find any self-hosted runners, with the bash code running on push then even syncing a fork gets me an email with errors in it.

File: .github/workflows/bash_unit_testing.yml (hunk 4 above)
Comment (daltonbohning) [reply .github/workflows/bash_unit_testing.yml:4]: What about when a new commit is pushed to this same branch/PR?

File: .github/workflows/bash_unit_testing.yml (hunk 5 above)
Comment (daltonbohning) [reply .github/workflows/bash_unit_testing.yml:17]: That sounds like we still fundamentally don't support PRs from forks then?

File: .github/workflows/bash_unit_testing.yml (hunk 4 above)
Comment (ashleypittman) [reply .github/workflows/bash_unit_testing.yml:4]: The line below catches that and is all any of our other actions use./

File: .github/workflows/bash_unit_testing.yml (hunk 5 above)
Comment (ashleypittman) [reply .github/workflows/bash_unit_testing.yml:17]: Not at all.  If there's a PR from a form to daos-stack then this would still trigger.  This change is for if there's a PR to a fork where the action would currently fail due to lack of self-hosted runners.

File: utils/utest.yaml (hunk 3 above)
Comment (daltonbohning) [reply utils/utest.yaml]: Sorry, I didn't mean you had to group those here, but that works! :)

Comment (ashleypittman) [reply utils/utest.yaml]: I did it mostly to see what the output looks like.  actions only appears to have one level of grouping.

**Review by daltonbohning — DISMISSED** [review]

File: src/control/run_go_tests.sh (hunk 2 above)
Comment (ashleypittman) [reply src/control/run_go_tests.sh]: I deleted this part of the change.  The intention wasn't ever to change the workflow, I just wanted to be able to get the gofmt results viewable outside of Intel CI systems.

**Review by mjmac — APPROVED** [review]

**Review by daltonbohning — APPROVED** [review]
